# NovaTel Customer Retention & Revenue Intelligence System

## Data Modeling

### Objective

Prepare an analysis-ready data model for SQL analysis, KPI development, and Power BI dashboard creation.

### Datasets

- telecom_customer_churn.csv
- telecom_zipcode_population.csv
- telecom_data_dictionary.csv (reference only)

### Goals

- Identify table roles
- Define relationships
- Prepare Power BI model design
- Validate join keys

In [1]:
import pandas as pd

churn_df = pd.read_csv("../data/raw/telecom_customer_churn.csv")

zipcode_df = pd.read_csv("../data/raw/telecom_zipcode_population.csv")

data_dict_df = pd.read_csv(
    "../data/raw/telecom_data_dictionary.csv",
    encoding="latin1"
)

In [2]:
print("Customer Churn:", churn_df.shape)
print("Zipcode Population:", zipcode_df.shape)
print("Data Dictionary:", data_dict_df.shape)

Customer Churn: (7043, 38)
Zipcode Population: (1671, 2)
Data Dictionary: (40, 3)


## Table Classification

### Purpose

Identify the role of each dataset within the analytical model.

Possible roles:

- Fact Table
- Dimension Table
- Reference Table

### Initial Modeling Assessment

#### Customer Churn

Expected Role:
- Fact Table

Reason:
- Contains customer-level records.
- Contains business events, revenue metrics, customer attributes, and churn outcomes.

#### Zipcode Population

Expected Role:
- Dimension Table

Reason:
- Contains descriptive geographic information.
- Can enrich customer records through Zip Code.

#### Data Dictionary

Expected Role:
- Reference Table

Reason:
- Used for documentation only.
- Does not participate in analytical relationships.

### Zip Code Key Validation

#### Purpose

Validate the Zip Code field in both datasets before creating a relationship.

Questions:

- How many unique zip codes exist in each dataset?
- Does the lookup table contain sufficient coverage?
- Is the Zip Code field suitable as a relationship key?

In [3]:
print("Customer Churn Unique Zip Codes:")
print(churn_df["Zip Code"].nunique())

print("\nPopulation Table Unique Zip Codes:")
print(zipcode_df["Zip Code"].nunique())

Customer Churn Unique Zip Codes:
1626

Population Table Unique Zip Codes:
1671


### Zip Code Key Findings

- Customer dataset contains 1,626 unique zip codes.
- Population dataset contains 1,671 unique zip codes.
- The population table provides broader geographic coverage than the customer dataset.
- Zip Code appears suitable as a relationship candidate.
- Join coverage validation is required before defining the relationship.

### Relationship Coverage Validation

#### Purpose

Verify that all customer zip codes exist in the population lookup table.

In [4]:
missing_zipcodes = set(churn_df["Zip Code"]) - set(zipcode_df["Zip Code"])

print("Missing Zip Codes:", len(missing_zipcodes))

Missing Zip Codes: 0


### Relationship Coverage Findings

- All customer zip codes were found in the population lookup table.
- No missing relationship keys were identified.
- No customer records will be lost during the join process.
- Zip Code is a valid relationship field for data modeling.

### Relationship Cardinality Validation

#### Purpose

Determine the relationship type between the customer dataset and the population lookup table.

In [5]:
zipcode_df["Zip Code"].duplicated().sum()

np.int64(0)

### Cardinality Findings

- No duplicate Zip Codes were found in the population lookup table.
- Each Zip Code appears only once in the lookup dataset.
- The relationship between the customer dataset and population dataset is One-to-Many.

Relationship:

Zipcode Population (One)
        ↓
Customer Churn (Many)

## Final Data Model

### Fact Table

#### Customer Churn

**Grain:**

- One row per customer

**Primary Key:**

- Customer ID

**Contains:**

- Customer demographics
- Geographic information
- Service information
- Contract information
- Revenue metrics
- Customer status information
- Churn information

---

### Dimension Table

#### Zipcode Population

**Primary Key:**

- Zip Code

**Contains:**

- Population information by zip code

**Purpose:**

- Geographic enrichment
- Regional analysis
- Population-based segmentation

---

### Reference Table

#### Data Dictionary

**Purpose:**

- Business definitions
- Field descriptions
- Documentation support

**Note:**

- Does not participate in model relationships

---

### Relationship

| From Table | Column | Relationship | To Table | Column |
|------------|---------|-------------|----------|---------|
| Zipcode Population | Zip Code | One-to-Many (1:*) | Customer Churn | Zip Code |

---

### Data Model Architecture

```text
                    Zipcode Population
                    ------------------
                    Zip Code (PK)
                    Population
                           │
                           │ 1 : *
                           ▼
                    Customer Churn
                    ------------------
                    Customer ID (PK)
                    Zip Code (FK)
                    Customer Attributes
                    Service Attributes
                    Revenue Metrics
                    Customer Status
                    Churn Information


Data Dictionary
------------------
Reference Only
(No Relationships)
```

---

### Model Summary

| Table | Role | Key |
|---------|---------|---------|
| Customer Churn | Fact Table | Customer ID |
| Zipcode Population | Dimension Table | Zip Code |
| Data Dictionary | Reference Table | N/A |

---

### Modeling Outcome

The NovaTel analytical model consists of:

- 1 Fact Table
- 1 Dimension Table
- 1 Reference Table

The model supports:

- Customer Analysis
- Churn Analysis
- Revenue Analysis
- Loyalty Analysis
- Product Analysis
- Geographic Analysis
- KPI Development
- Power BI Dashboard Development

The datasets are ready for Exploratory Data Analysis.

## Modeling Outcome

The NovaTel analytical model consists of:

- 1 Fact Table
- 1 Dimension Table
- 1 Reference Table

The model supports:

- Customer Analysis
- Churn Analysis
- Revenue Analysis
- Loyalty Analysis
- Geographic Analysis
- Power BI Dashboard Development

The datasets are ready for Exploratory Data Analysis.